doc/nsd.1
'\"
'\" The contents of this file are subject to the AOLserver Public License
'\" Version 1.1 (the "License"); you may not use this file except in
'\" compliance with the License. You may obtain a copy of the License at
'\" http://aolserver.com/.
'\"
'\" Software distributed under the License is distributed on an "AS IS"
'\" basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
'\" the License for the specific language governing rights and limitations
'\" under the License.
'\"
'\" The Original Code is AOLserver Code and related documentation
'\" distributed by AOL.
'\" 
'\" The Initial Developer of the Original Code is America Online,
'\" Inc. Portions created by AOL are Copyright (C) 1999 America Online,
'\" Inc. All Rights Reserved.
'\"
'\" Alternatively, the contents of this file may be used under the terms
'\" of the GNU General Public License (the "GPL"), in which case the
'\" provisions of GPL are applicable instead of those above.  If you wish
'\" to allow use of your version of this file only under the terms of the
'\" GPL and not to allow others to use your version of this file under the
'\" License, indicate your decision by deleting the provisions above and
'\" replace them with the notice and other provisions required by the GPL.
'\" If you do not delete the provisions above, a recipient may use your
'\" version of this file under either the License or the GPL.
'\" 
'\"
'\" $Header: /Users/dossy/Desktop/cvs/aolserver/doc/nsd.1,v 1.2 2002/09/28 19:58:27 jgdavidson Exp $
'\"
'\" 
.so man.macros
.TH nsd 1 "" AOLserver "AOLserver Applications"
.BS
'\" Note:  do not modify the .SH NAME line immediately below!
.SH NAME
nsd \- AOLserver program
.SH SYNOPSIS
\fBnsd\fR ?\fIarg ...\fR?
.BE

.SH DESCRIPTION
.PP
\fBNsd\fR is the AOLserver binary.

.SH "SEE ALSO"
tclsh(n)


ChangeLog
2004-06-14 20:28  Dossy Shiobara <dossy@panoptic.com>

	* nsd/connio.c: If nsend == 0, we would never call Ns_WriteConn to
	  flush the queued headers.  This could happen when we're sending
	  zero bytes of data as a response.  This fixes bug #971016 in the
	  case where fastpath.cache=false and fastpath.mmap=false and a
	  zero byte file is requested.

2004-06-14 20:25  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: unnecessary test for data != NULL actually caused
	  part of bug #971016 when fastpath.cache=false and
	  fastpath.mmap=true and the file requested is zero bytes, mmap()
	  returns 0 which gets passed along as data == NULL, causing
	  ReturnCharData() to not flush the queued headers.  removing the
	  if is safe as Ns_WriteConn will simply flush any queued data

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* Makefile: remove Makefile.module on distclean.
	* include/ns.h, nsd/log.c: Allow users to override logging functions.
	* nsd/str.c: Fix crash bug in Ns_Trim* when trimming NULL strings.

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* TAG aolserver_v40_r5

2004-03-09 13:35  Rob Crittenden <rcrittenden0569@aol.com>

	* nsd/tclhttp.c: The last header was being missed because the pointer
	  was being advanced past the \r\n.

2004-02-26 14:37  Dossy Shiobara <dossy@panoptic.com>

	* nsd/unix.c: uid_t is supposed to be unsigned, but apparently
	  isn't on OSX, apparently.

2004-02-25 19:56  Dossy Shiobara <dossy@panoptic.com>

	* nscgi/nscgi.c: Ns_ConnSetRequiredHeaders now requires length < 0
	  if you don't want a Content-Length header

2004-02-25 19:55  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: fix so that it's still possible to formulate a
	  response with no Content-Length header (such as for nscgi, or
	  HTTP streaming) where you don't know the Content-Length header in
	  advance

2004-02-24 07:15  Dossy Shiobara <dossy@panoptic.com>

	* nsd/: nsmain.c, nsd.h, driver.c: fix race condition where
	  DriverThreads are spawned but before they can Ns_SockListenEx,
	  the main thread does its NsClosePreBound and closes the prebound
	  sockets, making them unavailable to the DriverThreads if they
	  require root priv to listen on

2004-02-18 01:30  Dossy Shiobara <dossy@panoptic.com>

	* nsperm/nsperm.c: fixed bug #899364, restoring old nsperm
	  functionality as it was in 3.x.

2004-02-15  Jim Davidson <jgdavidson@aol.com>

	* nsd/conn.c:
	* include/ns.h:
	* nsd/return.c: Added new Ns_ConnGet/SetKeepAliveFlag APIs to
	allow custom request procedures to enable keep-alive directly.

	* nsd/driver.c: Minor updates to separate conn & driver structs.

2004-01-01  Scott S. Goodwin <scott@scottg.net>

        * nsd/adpparse.c: fixed bug #868600, where tag attribute values
	encapsulated within single quotes weren't being parsed properly.

2003-12-27  Scott S. Goodwin <scott@scottg.net>

        * nsd/urlopen.c: fixed bug #844809.

        * sample-config.tcl: fixed bug #805761.

2003-12-24  Scott S. Goodwin <scott@scottg.net>

	* include/nsthread.h: Bumped up NS_THREAD_NAMESIZE from 32 to 64.
	Mutex names were being truncated, making it impossible to distinguish
	mutexes with longer names when looking at stats.

        * nscgi/nscgi.c: Changed HTTPS env variable to be set to "on" or "off"
	instead of "1" or "0". There doesn't appear to be any standard for
	this, but Apache uses the on/off string in the environment.

2003-12-16  Scott S. Goodwin <scott@scottg.net>

        * nscgi/nscgi.c: Fixed issue with HTTPS environment variable not being
        set, and port not being set to 443 if connection is SSL'd. Manually
        applied these changes from a patch supplied by Vinod Kurup.

2003-11-16  Jim Davidson <jgdavidson@aol.com>

	* include/ns.h: Update version ids, added NS_SOCK_DROP and
	NS_TCL_TRACE constants and pre-queue function definitions.

	* nsd/nsd.h:  Added new "times" struct to Conn for more
	fine-grained timing.  Need to provide options to enable
	these time stamps plus ways to get the data.

	* nsd/driver.c:  Restructured in a few ways:

	1. Each loaded driver now has it's own thread.  This should
	allow some more concurrency (if necessary) and generally
	made the code more manageable.

	2. New pre-queue I/O callback API.  The API will be documented
	with an example HTTP module later.

	3. Conn allocation is handled by the driver threads and
	request buffers have been moved to the Conn from the Sock.

	4. Reads for synchrounous drivers (i.e., SSL) are handled
	by a pool of reader threads instead of by the conn threads.
	The number of threads is set with a "maxreaders" config
	option.  This change should enable both Host: header based
	virtual servers and pre-queue callbacks for SSL connections.

	5. New "maxsocks" config for drivers is now used to manage
	the concurrency limit of the server.  Conn's are now queued
	without limit into the connection pools and the previous
	"maxconns" virtual-server config is ignored.

	* nsd/queue.c: Cleaned up "connsperthread" code and updated
	NsQueueConn to support new driver code.

	* nsd/conn.c:
	* nsd/connio.c:
	* nsd/form.c:
	* nsd/nsd.h:
	* nsd/queue.c:
	* nsd/return.c: Moved connection request buffers from Sock
	to Conn structure.

	* nsd/server.c: Removed startup Conn pre-allocation as Conns
	are now allocated dynamically by the driver.  Also, removed
	the "maxconns" parameter which is no longer used.

	* nssock/nssock.c: Added new "async" option (default on)
	mostly to test new reader thread capabilities of the core
	driver code.

	* nsd/tclinit.c: Added a new generalized Tcl interp tracing
	facility and cleaned up some previous init code.  This
	should make it easier to support standard Tcl extensions
	and will be documented with an example module later.

	* nsd/nsmain.c: Fixed order of driver shutdown wait.

	* nsd/init.c: Fixed order of init calls.

	* nsd/binder.c:
	* nsd/dns.c: Named unnamed mutexes.

	* nsd/tclset.c:
	* nsd/tclsock.c: Removed unused functions.


nsd/nsd.tcl

# $Header: /Users/dossy/Desktop/cvs/aolserver/nsd/Attic/nsd.tcl,v 1.1 2000/10/09 20:29:53 kriston Exp $

#
# nsd.tcl --  The AOLserver Startup Script
#
#      This is a Tcl script that is sourced when AOLserver starts up.
#      A detailed reference is in "doc/config.txt".
#

ns_log notice "nsd.tcl: starting to read config file..."


#
# Set some Tcl variables that are commonly used throughout this file.
#

set httpport               8000
set httpsport              8443

# The hostname and address should be set to actual values.
set hostname               [ns_info hostname]
set address                [ns_info address]

set servername             "server1"
set serverdesc             "Server Name"

set homedir                [file dirname [ns_info config]]
set bindir                 [file dirname [ns_info nsd]]

set pageroot               ${homedir}/servers/${servername}/pages
set directoryfile          index.adp,index.html,index.htm

set ext [info sharedlibextension]

# nsssl: Only loads if keyfile.pem and certfile.pem exist.
#set sslmodule              nsssl${ext}  ;# Domestic 128-bit/1024-bit SSL.
set sslmodule              nsssle${ext} ;# Exportable 40-bit/512-bit SSL.
set sslkeyfile   ${homedir}/servers/${servername}/modules/nsssl/keyfile.pem
set sslcertfile  ${homedir}/servers/${servername}/modules/nsssl/certfile.pem

# nscp: Uncomment the sample password and log in with "nsadmin", password "x",
#       type "ns_crypt newpassword salt" and put the encrypted string below.
set nscp_port 9999
set nscp_addr 127.0.0.1
set nscp_user ""
#set nscp_user "nsadmin:t2GqvvaiIUbF2:" ;# sample user="nsadmin", pw="x".


#
# Global server parameters
#
ns_section "ns/parameters"
ns_param   home            $homedir
ns_param   debug           false


#
# Thread library (nsthread) parameters
#
ns_section "ns/threads"
ns_param   mutexmeter      true      ;# measure lock contention
#ns_param   stacksize [expr 128*1024] ;# Per-thread stack size.

#
# MIME types.
#
#  Note: AOLserver already has an exhaustive list of MIME types, but in
#  case something is missing you can add it here.
#
ns_section "ns/mimetypes"
ns_param   default         "*/*"     ;# MIME type for unknown extension.
ns_param   noextension     "*/*"     ;# MIME type for missing extension.
#ns_param   ".xls"          "application/vnd.ms-excel"


############################################################
#
# Server-level configuration
#
#  There is only one server in AOLserver, but this is helpful when multiple
#  servers share the same configuration file.  This file assumes that only
#  one server is in use so it is set at the top in the "server" Tcl variable.
#  Other host-specific values are set up above as Tcl variables, too.
#

ns_section "ns/servers"
ns_param   $servername     $serverdesc


#
# Server parameters
#
ns_section "ns/server/${servername}"
ns_param   directoryfile   $directoryfile
ns_param   pageroot        $pageroot
ns_param   globalstats     true      ;# Enable built-in statistics.
ns_param   urlstats        true      ;# Enable URL statistics.
ns_param   maxurlstats     1000      ;# Max number of URL's to do stats on.
ns_param   enabletclpages  false     ;# Parse *.tcl files in pageroot.

# Directory listings -- use an ADP or a Tcl proc to generate them.
#ns_param   directoryadp    $pageroot/dirlist.adp ;# Choose one or the other.
#ns_param   directoryproc   _ns_dirlist           ;#  ...but not both!
#ns_param   directorylisting simple               ;# Can be simple or fancy.


#
# ADP (AOLserver Dynamic Page) configuration
#
ns_section "ns/server/${servername}/adp"
ns_param   map             "/*.adp"  ;# Extensions to parse as ADP's.
#ns_param   map             "/*.html" ;# Any extension can be mapped.
ns_param   enableexpire    false     ;# Set "Expires: now" on all ADP's.
ns_param   enabledebug     false     ;# Allow Tclpro debugging with "?debug".

# ADP special pages
#ns_param   errorpage      ${pageroot}/errorpage.adp ;# ADP error page.


#
# ADP custom parsers -- see adp.c
#
ns_section "ns/server/${servername}/adp/parsers"
ns_param   adp             ".adp"    ;# adp is the default parser.


#
# Socket driver module (HTTP)  -- nssock
#
ns_section "ns/server/${servername}/module/nssock"
ns_param   port            $httpport
ns_param   hostname        $hostname
ns_param   address         $address


#
# Socket driver module (HTTPS) -- nsssl
#
#  nsssl does not load unless sslkeyfile/sslcertfile exist (above).
#
ns_section "ns/server/${servername}/module/nsssl"
ns_param   port            $httpsport
ns_param   hostname        $hostname
ns_param   address         $address
ns_param   keyfile         $sslkeyfile
ns_param   certfile        $sslcertfile


#
# Control port -- nscp
#
#  nscp does not load unless nscp_user is a valid user.
#
ns_section "ns/server/${servername}/module/nscp"
ns_param   port            $nscp_port
ns_param   address         $nscp_addr

ns_section "ns/server/${servername}/module/nscp/users"
ns_param   user            $nscp_user


#
# Access log -- nslog
#
ns_section "ns/server/${servername}/module/nslog"
ns_param   rolllog         true      ;# Should we roll log?
ns_param   rollonsignal    true      ;# Roll log on SIGHUP.
ns_param   rollhour        0         ;# Time to roll log.
ns_param   maxbackup       5         ;# Max number to keep around when rolling.


#
# CGI interface -- nscgi
#
#  WARNING: These directories must not live under pageroot.
#
ns_section "ns/server/${servername}/module/nscgi"
#ns_param   map "GET  /cgi /usr/local/cgi"     ;# CGI script file dir (GET).
#ns_param   map "POST /cgi /usr/local/cgi"     ;# CGI script file dir (POST).


#
# Modules to load
#
ns_section "ns/server/${servername}/modules"
ns_param   nssock          ${bindir}/nssock${ext}
ns_param   nslog           ${bindir}/nslog${ext}
#ns_param   nscgi           ${bindir}/nscgi${ext}  ;# Map the paths before using.
#ns_param   nsperm          ${bindir}/nsperm${ext} ;# Edit passwd before using.

#
# nsssl: Only loads if sslcertfile and sslkeyfile exist (see above).
#
if { [file exists $sslcertfile] && [file exists $sslkeyfile] } {
    ns_param nsssl ${bindir}/${sslmodule}
} else {
    ns_log warning "nsd.tcl: nsssl not loaded -- key/cert files do not exist."
}

#
# nscp: Only loads if nscp_user is set (see above).
#
if { $nscp_user != "" } {

    if ![string match "127.0.0.1" $nscp_addr] {
	# Anything but 127.0.0.1 is not recommended.
	ns_log warning "nsd.tcl: nscp listening on ${nscp_addr}:${nscp_port}"
    }
    ns_param nscp ${bindir}/nscp${ext}

} else {
    ns_log warning "nsd.tcl: nscp not loaded -- user/password is not set."
}

ns_log notice "nsd.tcl: finished reading config file."
